### ISSUE TITLE
Rethink argument order

### ISSUE THREAD

**Issue by gdalle** [JuliaDiff/DifferentiationInterface.jl#403]
Idea: allow `Vararg` for variable number of inputs `x = (a, b)`

Challenges:

- Avoid dispatch ambiguities
- Correctly handle seeds

**Move `extras` to the front**

- `extras` is mutated so it makes sense
- The last argument will always be `x`, which makes `Vararg` possible
- No ambiguities between `extras` and arguments

```julia
operator(f, [extras,] backend, x...)
operator(f!, y, [extras,] backend, x...)
operator!(f, result, [extras,] backend, x...)
operator!(f!, y, result, [extras,] backend, x...)
```

**Store seeds with their input**

Only allow `Tuple` arguments for `pushforward` and `hvp`

```julia
pushforward(f, backend, (a, da), (b, db))
```

If the tuples have more than two elements, it's batched seeds:

```julia
pushforward(f, backend, (a, (da1, da2)), (b, (db1, db2)))
```

For `pullback`, the only seed is on the output (and there is a single output), put it first:

```julia
pullback(f, backend, (dy,), a, b)
pullback(f, backend, (dy1, dy2), a, b)
```

**Comment by gdalle** [comment]
Alternative by @adrhill (code written by the user)

```julia
pushforward(f, backend, Inputs(; a=a, b=b), Seeds(; a=(da1, da2), b=(db1, db2))
```

**Comment by gdalle** [comment]
Compare putting the seeds with the inputs
```julia
pushforward(f, backend, InputWithSeed(a, da), InputWithSeed(b, db), Cache(c))
```
with separating them
```julia
pushforward(f, backend, Inputs(a, b, Cache(c)), Seeds(da, db, nothing))
```

**Comment by adrhill** [comment]
First of all, I see the need to move the `extras` to allow for `Vararg` inputs.

However, the nested tuples in the `pushforward` examples are hard to parse. I very much favor something the more explicit
> ```julia
> pushforward(f, backend, InputWithSeed(a, da), InputWithSeed(b, db), Cache(c))
> ```

This issue should be handled with future API changes from #311 in mind. In there, the following separation into activities [was proposed](https://github.com/gdalle/DifferentiationInterface.jl/issues/311#issuecomment-2228066950):

> * "input": care about initial value and derivatives, not mutated
> * "output": care about derivatives but not initial value, possibly mutated
> * "parameters": care about initial value but not derivatives, not mutated
> * "cache": don't care about initial value or derivatives, mutated

I wonder whether it's correct to distinguish between inputs and outputs here or whether activities apply to outputs as well. I'll propose the following:

| Name   | Initial value matters | Compute derivatives | Applicable to |
|:-------|:---------------------:|:-------------------:|:---:|
| `Active`  | Y[^1] | Y | input / output |
| `Context`/`Constant` | Y | N | input[^2] |
| `Auxiliary` | N | N | input / output |

[^1]: On outputs, initial value only matters for seeds in `pullback`.
[^2]: On outputs, `Auxiliary` should be used.

Seeds could be handled via MD, e.g. 
* `Active(x)`: input without seed
* `Active(x, dx)`: input with seed
* `Active(x, dx1, dx2)`: input with batched seeds

Using these more generic traits makes pullbacks more easily parsable and restores a little symmetry w.r.t. pushforwards:
 
```Julia
# pushforward
pushforward(f!, y, backend, Active(x, dx))
pushforward(f!, y, backend, Active(x, dx1, dx2))

# pullback
y = similar(dy)
pullback(f!, Active(y, dy), backend, xs...)
pullback(f!, Active(y, dy1, dy2), backend, xs...)
```

It could also [enable auxiliary outputs](https://github.com/gdalle/DifferentiationInterface.jl/issues/311#issuecomment-2269101734) (and possibly multi-output functions) in the future.

Outside of `pushforward` and `pullback`, which are advanced features, these types could all be optional/hidden, e.g.

```Julia
gradient(f!, y, backend, x1, x2, x3)
```
could fall back to 
```Julia
gradient(f!, Active(y), extras, backend, Active(x1), Active(x2), Active(x3))
```

**Comment by gdalle** [comment]
The symmetry between pushforward and pullback is not a meaningful objective: for pushforward we provide inputs and input seeds, for pullback we provide inputs and output seed. There's no going around it, even though you voluntarily picked the `f!` case for the pullback (I see what you did there). In the `f` case we don't provide `y`, only `dy`. In particular, we may not even know what the type of `y` will be.

**Comment by gdalle** [comment]
As for auxiliary outputs, no Julia AD package (that I know of) handles them. Typically, if we have `f(x) = (y, aux)`, to compute its Jacobian with ForwardDiff we would have to use `first \circ f`, and then call `f` a second time to compute `aux` anyway. I think the only solution would be to store this auxiliary data by mutating a cache. I'm not even sure we can trick Enzyme into avoiding the differentiation of `aux` if `f` returns `(y, aux)`.

**Comment by gdalle** [comment]
Another thing we haven't considered is that for in-place `pullback`, `da` and `db` both get updated in the following
```julia
pullback!(f, [extras, ] backend, dy, InputWithResult(a, da), InputWithResult(b, db))
```
Even worse, for in-place `hvp`, each input must carry both a seed and a result, to the tune of
```julia
hvp!(f, [extras, ] backend, InputWithSeedAndResult(a, da, dga), InputWithSeedAndResult(b, db, dgb))
```

**Comment by gdalle** [comment]
| operator      | seed       | result       |
| ------------- | ---------- | ------------ |
| `pushforward` | `da`, `db` | `dy`         |
| `pullback`    | `dy`       | `da`, `db`   |
| `hvp`         | `da`, `db` | `dga`, `dgb` |

**Comment by gdalle** [comment]
I just ran into a method ambiguity while working on #435. It's always the same story of confusion between generic fallbacks in the core package and specialized methods in the extensions.

```julia
  MethodError: value_and_derivative(::typeof(DifferentiationInterfaceTest.num_to_num), ::DifferentiationInterface.PushforwardDerivativeExtras{DifferentiationInterfaceForwardDiffExt.ForwardDiffOneArgPushforwardExtras{ForwardDiff.Tag{typeof(DifferentiationInterfaceTest.num_to_num), Float64}, ForwardDiff.Dual{ForwardDiff.Tag{typeof(DifferentiationInterfaceTest.num_to_num), Float64}, Float64, 1}}}, ::AutoForwardDiff{nothing, Nothing}, ::Float64) is ambiguous.
  
  Candidates:
    value_and_derivative(f!::F, y, ::AutoForwardDiff, x) where F
      @ DifferentiationInterfaceForwardDiffExt ~/Work/GitHub/Julia/DifferentiationInterface.jl/DifferentiationInterface/ext/DifferentiationInterfaceForwardDiffExt/twoarg.jl:97
    value_and_derivative(f::F, extras::DifferentiationInterface.PushforwardDerivativeExtras, backend::ADTypes.AbstractADType, x) where F
      @ DifferentiationInterface ~/Work/GitHub/Julia/DifferentiationInterface.jl/DifferentiationInterface/src/first_order/derivative.jl:74
  
  Possible fix, define
    value_and_derivative(::F, ::DifferentiationInterface.PushforwardDerivativeExtras, ::AutoForwardDiff, ::Any) where F
```
The backend implements `value_and_derivative` for `f!` but not `f`, and we can't dispatch on `y` not being an `Extras`.

**Comment by gdalle** [comment]
The `extras` argument has been moved by #435.
I don't think we're gonna store seeds with their input, or add `Active` annotation to the active arguments because there's probably gonna be only one. However we could add several `Constant`s or `Cache`s.

**Comment by gdalle** [comment]
This will be closed by #469, for now we stick to a single active argument

**Comment by gdalle** [comment]
Reopening because the design ideas remain interesting

**Comment by gdalle** [comment]
Closing because this will not change for a while
